testsuite/MDAnalysisTests/formats/test_libdcd.py

```diff
@@ -84,25 +84,48 @@ def dcd():
     with DCDFile(DCD) as dcd:
         yield dcd
 
+def _assert_compare_readers(old_reader, new_reader):
+    frame = old_reader.read()     # same as next(old_reader)
+    new_frame = new_reader.read() # same as next(new_reader)
 
-def test_pickle(dcd):
-    dcd.seek(len(dcd) - 1)
-    dump = pickle.dumps(dcd)
-    new_dcd = pickle.loads(dump)
+    assert old_reader.fname == new_reader.fname
+    assert old_reader.tell() == new_reader.tell()
+    assert_almost_equal(frame.xyz, new_frame.xyz)
+    assert_almost_equal(frame.unitcell, new_frame.unitcell)
 
-    assert dcd.fname == new_dcd.fname
-    assert dcd.tell() == new_dcd.tell()
+def test_pickle(dcd):
+    mid = len(dcd) // 2
+    dcd.seek(mid)
+    new_dcd = pickle.loads(pickle.dumps(dcd))
+    _assert_compare_readers(dcd, new_dcd)
 
+def test_pickle_last(dcd):
+    dcd.seek(len(dcd) - 1)
+    new_dcd = pickle.loads(pickle.dumps(dcd))
+    _assert_compare_readers(dcd, new_dcd)
 
 def test_pickle_closed(dcd):
     dcd.seek(len(dcd) - 1)
     dcd.close()
-    dump = pickle.dumps(dcd)
-    new_dcd = pickle.loads(dump)
+    new_dcd = pickle.loads(pickle.dumps(dcd))
 
     assert dcd.fname == new_dcd.fname
     assert dcd.tell() != new_dcd.tell()
 
+def test_pickle_after_read(dcd):
+    _ = dcd.read()
+    new_dcd = pickle.loads(pickle.dumps(dcd))
+    _assert_compare_readers(dcd, new_dcd)
+
+#@pytest.mark.xfail
+def test_pickle_immediately(dcd):
+    # do not seek before pickling: this seems to leave the DCDFile
+    # object in weird state: is this supposed to work?
+    new_dcd = pickle.loads(pickle.dumps(dcd))
```

Comment:
Pickle always fails if we have never done a `seek()` before attempting to pickle.
```
______________________________________ test_pickle_immediately _______________________________________

dcd = <MDAnalysis.lib.formats.libdcd.DCDFile object at 0x12a1e5890>

    def test_pickle_immediately(dcd):
        # do not seek before pickling: this seems to leave the DCDFile
        # object in weird state: is this supposed to work?
>       new_dcd = pickle.loads(pickle.dumps(dcd))

testsuite/MDAnalysisTests/formats/test_libdcd.py:124:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
MDAnalysis/lib/formats/libdcd.pyx:268: in MDAnalysis.lib.formats.libdcd.DCDFile.__setstate__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   OSError: DCD seek failed with DCD error=Normal EOF

MDAnalysis/lib/formats/libdcd.pyx:422: OSError
```

I find this behavior surprising (as a user) but maybe I don't understand if this is expected. But in that case we should raise an appropriate exception.

Reply:
This should work.

Reply:
Ok, thanks, then that's a bug.

Reply:
Except that it is only a bug in this PR... because it works with 1.0.0 and current develop
```
In [1]: import MDAnalysis as mda
In [2]: mda.__version__
Out[2]: '1.0.0'
In [3]: import MDAnalysis as mda; from MDAnalysis.tests import datafiles as data
In [4]: dcd = mda.lib.formats.libdcd.DCDFile(data.COORDINATES_DCD)
In [5]: import pickle
In [7]: pickle.dumps(data.COORDINATES_DCD)
Out[7]: b'\x80\x03Xg\x00\x00\x00~/anaconda3/envs/mda3/lib/python3.6/site-packages/MDAnalysisTests/data/coordinates/test.dcdq\x00.'
```
and also works for trr
```
In [8]: trr = mda.lib.formats.libmdaxdr.TRRFile(data.COORDINATES_TRR)
In [9]: pickle.dumps(trr)
Out[9]: b'\x80\x03cMDAnalysis.lib.formats.libmdaxdr\nTRRFile\nq\x00Xg\x00\x00\x00~/anaconda3/envs/mda3/lib/python3.6/site-packages/MDAnalysisTests/data/coordinates/test.trrq\x01X\x01\x00\x00\x00rq\x02\x86q\x03Rq\x04(K\x01K\x00NK\x00tq\x05b.'
```

Reply:
The approach I took here is flawed (so it also shouldn't work in current develop--when `pickle.loading`).

Reply:
Hm, I need to check what commit I used for this:
```
In [13]: dcd = mda.lib.formats.libdcd.DCDFile(data.COORDINATES_DCD)

In [14]: pickle.dumps(dcd)
Out[14]: b'\x80\x03cMDAnalysis.lib.formats.libdcd\nDCDFile\nq\x00Xj\x00\x00\x00~/anaconda3/envs/mda3dev/lib/python3.7/site-packages/MDAnalysisTests/data/coordinates/test.dcdq\x01X\x01\x00\x00\x00rq\x02\x86q\x03Rq\x04K\x01K\x00\x86q\x05b.'

In [15]: mda.__version__
Out[15]: '2.0.0-dev0'
```

Reply:
Oddly enough, `data.COORDINATES_DCD` (which is used in the testsuite) fails at pickling while `DCD` fails at reading:
```python
>>> dcd = mda.lib.formats.libdcd.DCDFile(data.COORDINATES_DCD)
>>> read_p = pickle.loads(pickle.dumps(dcd))
>>> read_p.read()
OSError: Reading DCD header failed: format of DCD file is wrong

```